// @strapi/upload/dist/admin/future/services/assets.mjs
import { uploadApi } from './api.mjs';

const assetsApi = uploadApi.injectEndpoints({
    endpoints: (builder)=>({
            getAssets: builder.query({
                query: (params = {})=>{
                    const { folder, ...rest } = params;
                    const queryParams = {
                        ...rest
                    };
                    if (folder != null) {
                        queryParams['filters'] = {
                            $and: [
                                {
                                    folder: {
                                        id: folder
                                    }
                                }
                            ]
                        };
                    } else {
                        queryParams['filters'] = {
                            $and: [
                                {
                                    folder: {
                                        id: {
                                            $null: true
                                        }
                                    }
                                }
                            ]
                        };
                    }
                    return {
                        url: '/upload/files',
                        method: 'GET',
                        config: {
                            params: queryParams
                        }
                    };
                },
                transformResponse: (response)=>response,
                providesTags: (result)=>result ? [
                        ...result.results.map(({ id })=>({
                                type: 'Asset',
                                id
                            })),
                        {
                            type: 'Asset',
                            id: 'LIST'
                        }
                    ] : [
                        {
                            type: 'Asset',
                            id: 'LIST'
                        }
                    ]
            }),
            getAsset: builder.query({
                query: (id)=>({
                        url: `/upload/files/${id}`,
                        method: 'GET'
                    }),
                providesTags: (_result, _error, id)=>[
                        {
                            type: 'Asset',
                            id
                        }
                    ]
            }),
            /**
     * Update the editable metadata of an existing asset.
     * Hits the legacy `POST /upload?id=<id>` endpoint which dispatches to
     * `admin-upload.updateFileInfo`.
     */ updateAsset: builder.mutation({
                query: ({ id, fileInfo })=>{
                    const formData = new FormData();
                    formData.append('fileInfo', JSON.stringify(fileInfo));
                    return {
                        url: '/upload',
                        method: 'POST',
                        data: formData,
                        config: {
                            params: {
                                id
                            }
                        }
                    };
                },
                invalidatesTags: (_result, _error, { id })=>[
                        {
                            type: 'Asset',
                            id
                        },
                        {
                            type: 'Asset',
                            id: 'LIST'
                        }
                    ]
            })
        })
});
const { useGetAssetsQuery, useGetAssetQuery, useUpdateAssetMutation } = assetsApi;

export { useGetAssetQuery, useGetAssetsQuery, useUpdateAssetMutation };
