// @strapi/database/dist/migrations/internal.js
'use strict';

var common = require('./common.js');
var index = require('./internal-migrations/index.js');
var storage = require('./storage.js');
var logger = require('./logger.js');

const createInternalMigrationProvider = (db)=>{
    const context = {
        db
    };
    const migrations = [
        ...index.internalMigrations
    ];
    // Lazy: defer `umzug` (and its inquirer / @rushstack chain) until first call
    let lazyProvider;
    const provider = ()=>{
        if (lazyProvider) return lazyProvider;
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        const { Umzug: UmzugCtor } = require('umzug');
        lazyProvider = new UmzugCtor({
            storage: storage.createStorage({
                db,
                tableName: 'strapi_migrations_internal'
            }),
            logger: {
                info (message) {
                    // NOTE: only log internal migration in debug mode
                    db.logger.debug(logger.transformLogMessage('info', message));
                },
                warn (message) {
                    db.logger.warn(logger.transformLogMessage('warn', message));
                },
                error (message) {
                    db.logger.error(logger.transformLogMessage('error', message));
                },
                debug (message) {
                    db.logger.debug(logger.transformLogMessage('debug', message));
                }
            },
            context,
            migrations: ()=>migrations.map((migration)=>{
                    return {
                        name: migration.name,
                        up: common.wrapTransaction(context.db)(migration.up),
                        down: common.wrapTransaction(context.db)(migration.down)
                    };
                })
        });
        return lazyProvider;
    };
    return {
        async register (migration) {
            migrations.push(migration);
        },
        async shouldRun () {
            const pendingMigrations = await provider().pending();
            return pendingMigrations.length > 0;
        },
        async up () {
            await provider().up();
        },
        async down () {
            await provider().down();
        }
    };
};

exports.createInternalMigrationProvider = createInternalMigrationProvider;
