// @strapi/upload/dist/server/services/ai-metadata-jobs.js
'use strict';

var aiMetadataJob = require('../models/ai-metadata-job.js');

const createAIMetadataJobsService = ({ strapi })=>({
        async createJob () {
            const job = await strapi.db.query(aiMetadataJob.AI_METADATA_JOB_UID).create({
                data: {
                    status: 'processing',
                    createdAt: new Date()
                }
            });
            return job.id;
        },
        async getJob (jobId) {
            return strapi.db.query(aiMetadataJob.AI_METADATA_JOB_UID).findOne({
                where: {
                    id: jobId
                }
            });
        },
        async updateJob (jobId, updates) {
            await strapi.db.query(aiMetadataJob.AI_METADATA_JOB_UID).update({
                where: {
                    id: jobId
                },
                data: updates
            });
        },
        async deleteJob (jobId) {
            await strapi.db.query(aiMetadataJob.AI_METADATA_JOB_UID).delete({
                where: {
                    id: jobId
                }
            });
        },
        async getLatestActiveJob () {
            // Return the most recent job, regardless of status
            // This allows the frontend to see completed/failed status
            return strapi.db.query(aiMetadataJob.AI_METADATA_JOB_UID).findOne({
                orderBy: {
                    createdAt: 'desc'
                }
            });
        }
    });

exports.createAIMetadataJobsService = createAIMetadataJobsService;
