// @strapi/upload/dist/admin/future/pages/Assets/components/AssetDetails/AssetDetailsDrawer.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var strapiAdmin = require('@strapi/admin/strapi-admin');
var designSystem = require('@strapi/design-system');
var icons = require('@strapi/icons');
var reactIntl = require('react-intl');
var styledComponents = require('styled-components');
var Drawer = require('../../../../components/Drawer.js');
var enums = require('../../../../enums.js');
var assets = require('../../../../services/assets.js');
var folders = require('../../../../services/folders.js');
var files = require('../../../../utils/files.js');
var getAssetIcon = require('../../../../utils/getAssetIcon.js');
var translations = require('../../../../utils/translations.js');
var AssetPreview = require('./AssetPreview.js');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);

// Name of the parameter to look for in the URL to open the drawer
const URL_PARAM = 'assetId';
/* -------------------------------------------------------------------------------------------------
 * useAssetDetailsParam - sync drawer visibility with URL ?{URL_PARAM}={id}
 * -----------------------------------------------------------------------------------------------*/ const useAssetDetailsParam = ()=>{
    const [{ query }, setQuery] = strapiAdmin.useQueryParams();
    const detailsId = query?.[URL_PARAM];
    const assetId = detailsId ? parseInt(detailsId, 10) : null;
    const hasValidId = assetId !== null && !Number.isNaN(assetId);
    // Closing is driven by removing the URL param (a navigation), so navigation
    // guards like <Blocker> can intercept it. `isMounted` keeps the drawer in the
    // tree through the slide-out: it stays true once opened and only flips false
    // when the close animation actually ends (see onCloseAnimationEnd), so the
    // close duration lives entirely in CSS — no JS timer.
    const [isMounted, setIsMounted] = React__namespace.useState(hasValidId);
    const displayAssetId = React__namespace.useRef(null);
    React__namespace.useEffect(()=>{
        if (hasValidId) {
            displayAssetId.current = assetId;
            setIsMounted(true);
        }
    }, [
        hasValidId,
        assetId
    ]);
    const onCloseAnimationEnd = React__namespace.useCallback((event)=>{
        // Ignore animations bubbling up from descendants, and the slide-in.
        if (event.target === event.currentTarget && !hasValidId) {
            setIsMounted(false);
        }
    }, [
        hasValidId
    ]);
    const openDetails = React__namespace.useCallback((id)=>{
        setQuery({
            [URL_PARAM]: String(id)
        }, 'push', true);
    }, [
        setQuery
    ]);
    const closeDetails = React__namespace.useCallback(()=>{
        setQuery({
            [URL_PARAM]: undefined
        }, 'remove', true);
    }, [
        setQuery
    ]);
    return {
        assetId: hasValidId ? assetId : displayAssetId.current,
        isVisible: hasValidId,
        shouldRenderDrawer: isMounted,
        onCloseAnimationEnd,
        openDetails,
        closeDetails
    };
};
const DetailItemContainer = styledComponents.styled(designSystem.Flex)`
  flex: 0 0 calc(50% - ${({ theme })=>theme.spaces[2]});
`;
const DetailItem = ({ label, value })=>/*#__PURE__*/ jsxRuntime.jsxs(DetailItemContainer, {
        direction: "column",
        justifyContent: "flex-start",
        alignItems: "flex-start",
        gap: 1,
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                variant: "sigma",
                textColor: "neutral600",
                fontWeight: "semiBold",
                textTransform: "uppercase",
                children: label
            }),
            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                variant: "pi",
                textColor: "neutral700",
                children: value ?? '-'
            })
        ]
    });
/* -------------------------------------------------------------------------------------------------
 * DetailField
 * -----------------------------------------------------------------------------------------------*/ const StyledWarning = styledComponents.styled(icons.WarningCircle)`
  width: 1.6rem;
  height: 1.6rem;

  path {
    fill: ${({ theme })=>theme.colors.warning500};
  }
`;
const DetailField = ({ name, label, required })=>{
    const { formatMessage } = reactIntl.useIntl();
    const field = strapiAdmin.useField(name);
    const isSubmitting = strapiAdmin.useForm('DetailField', (state)=>state.isSubmitting);
    const value = field.value ?? '';
    const emptyTooltipLabel = formatMessage({
        id: translations.getTranslationKey('asset-details.field.empty'),
        defaultMessage: '{label} is currently empty.'
    }, {
        label
    });
    return /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Field.Root, {
        name: name,
        required: required,
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Field.Label, {
                children: label
            }),
            /*#__PURE__*/ jsxRuntime.jsx(designSystem.TextInput, {
                value: value,
                onChange: (event)=>field.onChange(name, event.target.value),
                endAction: !value ? /*#__PURE__*/ jsxRuntime.jsx(designSystem.Tooltip, {
                    label: emptyTooltipLabel,
                    children: /*#__PURE__*/ jsxRuntime.jsx(StyledWarning, {
                        "aria-label": emptyTooltipLabel,
                        role: "img"
                    })
                }) : undefined,
                type: "text",
                disabled: isSubmitting
            })
        ]
    });
};
const LocationField = ({ label, rootLabel, folders })=>{
    const field = strapiAdmin.useField('folder');
    const isSubmitting = strapiAdmin.useForm('LocationField', (state)=>state.isSubmitting);
    return /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Field.Root, {
        name: "folder",
        required: true,
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Field.Label, {
                children: label
            }),
            /*#__PURE__*/ jsxRuntime.jsxs(designSystem.SingleSelect, {
                value: field.value == null ? '' : String(field.value),
                onChange: (value)=>{
                    const next = value === '' ? null : Number(value);
                    field.onChange('folder', next);
                },
                disabled: isSubmitting,
                children: [
                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.SingleSelectOption, {
                        value: "",
                        children: rootLabel
                    }),
                    folders.map((folder)=>/*#__PURE__*/ jsxRuntime.jsx(designSystem.SingleSelectOption, {
                            value: String(folder.id),
                            children: folder.name
                        }, folder.id))
                ]
            })
        ]
    });
};
const AssetDetails = ({ asset })=>{
    const { formatMessage, formatDate } = reactIntl.useIntl();
    const { toggleNotification } = strapiAdmin.useNotification();
    const { data: folders$1 = [] } = folders.useGetAllFoldersQuery();
    const [updateAsset] = assets.useUpdateAssetMutation();
    const isImage = asset.mime?.includes(enums.AssetType.Image);
    const initialValues = {
        name: asset.name ?? '',
        caption: asset.caption ?? '',
        alternativeText: asset.alternativeText ?? '',
        folder: typeof asset.folder === 'object' && asset.folder !== null ? asset.folder.id ?? null : asset.folder ?? null
    };
    const handleSubmit = async (values)=>{
        const res = await updateAsset({
            id: asset.id,
            fileInfo: {
                name: values.name,
                caption: values.caption,
                alternativeText: values.alternativeText,
                folder: values.folder
            }
        });
        if ('error' in res) {
            toggleNotification({
                type: 'danger',
                message: formatMessage({
                    id: translations.getTranslationKey('asset-details.update.error'),
                    defaultMessage: 'Failed to update the file.'
                })
            });
            return;
        }
        toggleNotification({
            type: 'success',
            message: formatMessage({
                id: translations.getTranslationKey('asset-details.update.success'),
                defaultMessage: 'File updated'
            })
        });
    };
    return(// `key={asset.id}` resets the form when the drawer switches to a different
    // asset so cached values from the previous asset don't bleed in.
    /*#__PURE__*/ jsxRuntime.jsx(strapiAdmin.Form, {
        method: "POST",
        initialValues: initialValues,
        onSubmit: handleSubmit,
        children: ({ modified, isSubmitting, values, resetForm })=>{
            const nameIsEmpty = (values.name ?? '').trim() === '';
            return /*#__PURE__*/ jsxRuntime.jsxs(jsxRuntime.Fragment, {
                children: [
                    /*#__PURE__*/ jsxRuntime.jsx(strapiAdmin.Blocker, {
                        onProceed: resetForm
                    }),
                    /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                        direction: "column",
                        alignItems: "stretch",
                        gap: 4,
                        paddingTop: 4,
                        paddingBottom: 4,
                        paddingLeft: 5,
                        paddingRight: 5,
                        children: [
                            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                                variant: "beta",
                                fontWeight: "semiBold",
                                tag: "h3",
                                children: formatMessage({
                                    id: translations.getTranslationKey('asset-details.fileInfo'),
                                    defaultMessage: 'File info'
                                })
                            }),
                            /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                                wrap: "wrap",
                                gap: 4,
                                background: "neutral100",
                                paddingTop: 4,
                                paddingBottom: 4,
                                paddingLeft: 6,
                                paddingRight: 6,
                                alignItems: "flex-start",
                                children: [
                                    /*#__PURE__*/ jsxRuntime.jsx(DetailItem, {
                                        label: formatMessage({
                                            id: translations.getTranslationKey('asset-details.creationDate'),
                                            defaultMessage: 'Creation date'
                                        }),
                                        value: asset.createdAt ? formatDate(new Date(asset.createdAt), {
                                            dateStyle: 'long',
                                            timeStyle: 'short'
                                        }) : null
                                    }),
                                    /*#__PURE__*/ jsxRuntime.jsx(DetailItem, {
                                        label: formatMessage({
                                            id: translations.getTranslationKey('asset-details.lastUpdated'),
                                            defaultMessage: 'Last updated'
                                        }),
                                        value: asset.updatedAt ? formatDate(new Date(asset.updatedAt), {
                                            dateStyle: 'long',
                                            timeStyle: 'short'
                                        }) : null
                                    }),
                                    /*#__PURE__*/ jsxRuntime.jsx(DetailItem, {
                                        label: formatMessage({
                                            id: translations.getTranslationKey('asset-details.createdBy'),
                                            defaultMessage: 'Created by'
                                        }),
                                        value: asset.createdBy ? strapiAdmin.getDisplayName({
                                            firstname: asset.createdBy.firstname ?? undefined,
                                            lastname: asset.createdBy.lastname ?? undefined,
                                            username: asset.createdBy.username ?? undefined,
                                            email: asset.createdBy.email ?? undefined
                                        }) ?? '-' : null
                                    }),
                                    /*#__PURE__*/ jsxRuntime.jsx(DetailItem, {
                                        label: formatMessage({
                                            id: translations.getTranslationKey('asset-details.size'),
                                            defaultMessage: 'Size'
                                        }),
                                        value: asset.size ? files.formatBytes(asset.size, 1) : null
                                    }),
                                    isImage && (asset.width != null || asset.height != null) && /*#__PURE__*/ jsxRuntime.jsx(DetailItem, {
                                        label: formatMessage({
                                            id: translations.getTranslationKey('asset-details.dimensions'),
                                            defaultMessage: 'Dimensions'
                                        }),
                                        value: asset.width != null && asset.height != null ? `${asset.width} × ${asset.height}` : null
                                    }),
                                    /*#__PURE__*/ jsxRuntime.jsx(DetailItem, {
                                        label: formatMessage({
                                            id: translations.getTranslationKey('asset-details.extension'),
                                            defaultMessage: 'Extension'
                                        }),
                                        value: files.getFileExtension(asset.ext)
                                    }),
                                    /*#__PURE__*/ jsxRuntime.jsx(DetailItem, {
                                        label: formatMessage({
                                            id: translations.getTranslationKey('asset-details.assetId'),
                                            defaultMessage: 'Asset ID'
                                        }),
                                        value: String(asset.id)
                                    })
                                ]
                            }),
                            /*#__PURE__*/ jsxRuntime.jsx(DetailField, {
                                name: "name",
                                label: formatMessage({
                                    id: translations.getTranslationKey('asset-details.fileName'),
                                    defaultMessage: 'File name'
                                }),
                                required: true
                            }),
                            /*#__PURE__*/ jsxRuntime.jsx(LocationField, {
                                label: formatMessage({
                                    id: translations.getTranslationKey('asset-details.location'),
                                    defaultMessage: 'Location'
                                }),
                                rootLabel: formatMessage({
                                    id: translations.getTranslationKey('plugin.home'),
                                    defaultMessage: 'Home'
                                }),
                                folders: folders$1
                            }),
                            isImage && /*#__PURE__*/ jsxRuntime.jsxs(jsxRuntime.Fragment, {
                                children: [
                                    /*#__PURE__*/ jsxRuntime.jsx(DetailField, {
                                        name: "caption",
                                        label: formatMessage({
                                            id: translations.getTranslationKey('asset-details.caption'),
                                            defaultMessage: 'Caption'
                                        })
                                    }),
                                    /*#__PURE__*/ jsxRuntime.jsx(DetailField, {
                                        name: "alternativeText",
                                        label: formatMessage({
                                            id: translations.getTranslationKey('asset-details.alternativeText'),
                                            defaultMessage: 'Alternative text'
                                        })
                                    })
                                ]
                            }),
                            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
                                justifyContent: "flex-end",
                                gap: 2,
                                paddingTop: 2,
                                children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Button, {
                                    type: "submit",
                                    variant: "default",
                                    loading: isSubmitting,
                                    // File name is required; block submit when it's empty or whitespace so the API can't 400 on a blank value.
                                    disabled: !modified || isSubmitting || nameIsEmpty,
                                    children: formatMessage({
                                        id: translations.getTranslationKey('asset-details.save'),
                                        defaultMessage: 'Save changes'
                                    })
                                })
                            })
                        ]
                    })
                ]
            });
        }
    }, asset.id));
};
const DrawerHeader = ({ asset, closeDetails })=>{
    const DocIcon = asset ? getAssetIcon.getAssetIcon(asset.mime, asset.ext) : icons.FileError;
    return /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
        gap: 2,
        paddingLeft: 5,
        paddingTop: 3,
        paddingBottom: 3,
        paddingRight: 3,
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(DocIcon, {
                width: 20,
                height: 20
            }),
            /*#__PURE__*/ jsxRuntime.jsx(Drawer.Drawer.Title, {
                asChild: true,
                children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                    variant: "omega",
                    fontWeight: "semiBold",
                    overflow: "hidden",
                    ellipsis: true,
                    tag: "h2",
                    children: asset.name
                })
            }),
            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
                marginLeft: "auto",
                children: /*#__PURE__*/ jsxRuntime.jsx(Drawer.Drawer.CloseButton, {
                    onClose: closeDetails,
                    children: /*#__PURE__*/ jsxRuntime.jsx(icons.ArrowLineRight, {})
                })
            })
        ]
    });
};
const DrawerContent = ({ assetId, closeDetails })=>{
    const { formatMessage } = reactIntl.useIntl();
    const { data: asset, isLoading, error } = assets.useGetAssetQuery(assetId, {
        refetchOnMountOrArgChange: false,
        refetchOnReconnect: false,
        refetchOnFocus: false
    });
    if (isLoading) {
        return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
            justifyContent: "center",
            padding: 8,
            children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Loader, {
                children: formatMessage({
                    id: 'app.loading',
                    defaultMessage: 'Loading...'
                })
            })
        });
    }
    if (error || !asset) {
        return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Flex, {
            direction: "column",
            alignItems: "stretch",
            gap: 4,
            padding: 4,
            children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Alert, {
                variant: "danger",
                closeLabel: formatMessage({
                    id: 'global.close',
                    defaultMessage: 'Close'
                }),
                onClose: closeDetails,
                children: formatMessage({
                    id: translations.getTranslationKey('asset-details.error'),
                    defaultMessage: 'Failed to load file details.'
                })
            })
        });
    }
    return /*#__PURE__*/ jsxRuntime.jsxs(jsxRuntime.Fragment, {
        children: [
            /*#__PURE__*/ jsxRuntime.jsx(DrawerHeader, {
                asset: asset,
                closeDetails: closeDetails
            }),
            /*#__PURE__*/ jsxRuntime.jsxs(Drawer.Drawer.ScrollableContent, {
                children: [
                    /*#__PURE__*/ jsxRuntime.jsx(AssetPreview.AssetPreview, {
                        asset: asset
                    }),
                    /*#__PURE__*/ jsxRuntime.jsx(AssetDetails, {
                        asset: asset
                    })
                ]
            })
        ]
    });
};
/* -------------------------------------------------------------------------------------------------
 * AssetDetailsDrawer
 * -----------------------------------------------------------------------------------------------*/ const AssetDetailsDrawer = ()=>{
    const { formatMessage } = reactIntl.useIntl();
    const { assetId, isVisible, shouldRenderDrawer, onCloseAnimationEnd, closeDetails } = useAssetDetailsParam();
    if (!shouldRenderDrawer || assetId === null) {
        return null;
    }
    return /*#__PURE__*/ jsxRuntime.jsxs(Drawer.Drawer.Root, {
        isVisible: isVisible,
        onClose: closeDetails,
        children: [
            /*#__PURE__*/ jsxRuntime.jsx("div", {
                children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.VisuallyHidden, {
                    children: [
                        /*#__PURE__*/ jsxRuntime.jsx(Drawer.Drawer.Title, {
                            children: formatMessage({
                                id: translations.getTranslationKey('asset-details.title'),
                                defaultMessage: 'File details'
                            })
                        }),
                        /*#__PURE__*/ jsxRuntime.jsx(Drawer.Drawer.Description, {
                            children: formatMessage({
                                id: translations.getTranslationKey('asset-details.description'),
                                defaultMessage: 'Displays file information and metadata'
                            })
                        })
                    ]
                })
            }),
            /*#__PURE__*/ jsxRuntime.jsx(Drawer.Drawer.Body, {
                animationDirection: "left",
                width: "41.6rem",
                height: "100vh",
                onAnimationEnd: onCloseAnimationEnd,
                children: /*#__PURE__*/ jsxRuntime.jsx(DrawerContent, {
                    assetId: assetId,
                    closeDetails: closeDetails
                })
            })
        ]
    });
};

exports.AssetDetails = AssetDetails;
exports.AssetDetailsDrawer = AssetDetailsDrawer;
exports.useAssetDetailsParam = useAssetDetailsParam;
